feat: create `SubscriptExpr` class

apkrelling opened the pull request:
## Pull Request description

This PR adds `SubscriptExpr` class. 
This PR is an attempt to resolve #98   .

- [X] created NoneType in types/base.py
- [X] created LiteralNone in literals.base.py -> mypy has an issue here
- [X] added NoneType in types/__init__.py
- [X] added LiteralNone in literals/__init__.py 
- [X] added new ASTKind
- [X] created `SubscriptExpr` class in the new `astx/subscript.py` file
- [X] updated `__init__.py` to include the new imports and classes in the `__all__` list
- [ ] added test for `SubscriptExpr` class
- [ ] added transpiler visit method for `SubscriptExpr`
- [ ] added transpiler test for `SubscriptExpr`


## How to test these changes

```python
from astx.subscript import SubscriptExpr
from astx.variables import Variable
from astx.literals import LiteralInt32

# Variables
a_var = Variable(name="a")

# SubscriptExpr
subscr_expr = SubscriptExpr(
    var_=a_var,
    index=LiteralInt32(0),
    upper=LiteralInt32(10),
    step=LiteralInt32(2),
)

# Print the ast representation
subscr_expr
```
```python
from astx.subscript import SubscriptExpr
from astx.variables import Variable
from astx.literals import LiteralInt32

# Variables
a_var = Variable(name="a")

# SubscriptExpr
subscr_expr = SubscriptExpr(
    var_=a_var,
    index=LiteralInt32(0),
)

subscr_expr
```
outputs:
![subs_upper](https://github.com/user-attachments/assets/a1e20bdd-c51a-4025-ae73-0ebb8d0ba98e)
![subs_index](https://github.com/user-attachments/assets/e54e7436-1692-46e9-9d38-314cd66604e9)


## Pull Request checklists

This PR is a:

- [ ] bug-fix
- [X] new feature
- [ ] maintenance

About this PR:

- [ ] it includes tests.
- [ ] the tests are executed on CI.
- [ ] the tests generate log file(s) (path).
- [ ] pre-commit hooks were executed locally.
- [ ] this PR requires a project documentation update.

Author's checklist:

- [ ] I have reviewed the changes and it contains no misspelling.
- [ ] The code is well commented, especially in the parts that contain more
      complexity.
- [ ] New and old tests passed locally.

## Additional information

TO-DO:
* waiting for approval of the work done so far before doing transpiler method ad tests
* fix `mypy` probem with `LiteralNone` in `literals.base.py`

xmnlab commented:
just for the record:

```python
>>> import ast
>>> ast.dump(ast.parse("a[0]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Constant(value=0), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0:1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Slice(lower=Constant(value=0), upper=Constant(value=1)), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0:1:1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Slice(lower=Constant(value=0), upper=Constant(value=1), step=Constant(value=1)), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0,1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Tuple(elts=[Constant(value=0), Constant(value=1)], ctx=Load()), ctx=Load()))], type_ignores=[])" 
```

xmnlab commented:
just for the record, some python refs:

```python
>>> import ast
>>> ast.dump(ast.parse("a[0]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Constant(value=0), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0:1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Slice(lower=Constant(value=0), upper=Constant(value=1)), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0:1:1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Slice(lower=Constant(value=0), upper=Constant(value=1), step=Constant(value=1)), ctx=Load()))], type_ignores=[])"
>>> ast.dump(ast.parse("a[0,1]"))
"Module(body=[Expr(value=Subscript(value=Name(id='a', ctx=Load()), slice=Tuple(elts=[Constant(value=0), Constant(value=1)], ctx=Load()), ctx=Load()))], type_ignores=[])"
>>> 
Module(body=[Expr(value=Subscript(value=Call(func=Name(id='func', ctx=Load()), args=[], keywords=[]), slice=Slice(lower=Constant(value=0), upper=Constant(value=1), step=Constant(value=1)), ctx=Load()))], type_ignores=[])"
```

xmnlab commented:
@apkrelling considering your previous example, and considering the python subscript, I would suggest some alternatives:


```
# a_var[0]
subscr_expr = SubscriptExpr(
    value=a_var,
    index=LiteralInt32(0),
)

# a_var[0:10]
subscr_expr = SubscriptExpr(
    value=a_var,
    lower=LiteralInt32(0),
    upper=LiteralInt32(10),
)

# a_var[0:10:2]
subscr_expr = SubscriptExpr(
    value=a_var,
    lower=LiteralInt32(0),
    upper=LiteralInt32(10),
    step=LiteralInt32(2),
)
```

for now we don't need to implement the `a[:,:,:]` structure. and maybe we would need to expand it to use something like slices to have something more flexible. but for now this would be good enough

let me know if that makes sense please.

xmnlab commented on src/astx/subscript.py:
```diff
@@ -0,0 +1,97 @@
+
+        Parameters
+        ----------
+        - value: The expression representing the object being indexed (e.g.,
```
use the numpydocs format: https://numpydoc.readthedocs.io/en/latest/format.html

apkrelling commented:
- [X] added test for `SubscriptExpr` class
- [X] added transpiler visit method for `SubscriptExpr`
- [X] added transpiler test for `SubscriptExpr`

TO-DO:
- fix `mypy` probem in with `LiteralNone` in `literals.base.py`
- fix `mypy` probem in with `self.index.get_struct(simplified)` in `subscript.py`

apkrelling commented:
Also, here are the AST representations in a jupyter notebook:
![subs1_ipynb](https://github.com/user-attachments/assets/1b7c1ffb-bda1-4764-a998-d163d9d58fc4)
![subs2_ipynb](https://github.com/user-attachments/assets/ceee60b3-65f8-44af-b4a4-f1ef191162af)

xmnlab reviewed (commented):
@apkrelling , it looks good to me.
just fix the linter issues, and it should be good to go.
thanks, appreciate that.

apkrelling commented:
The `mypy` issues were fixed. just please note that I used the work "indexed" for the "value" (the array being indexed) in the AST representation (image below). I thought it would make more sense, but please let me know if you disagree. If that's the case, I can change it and leave as "value".
![indexed](https://github.com/user-attachments/assets/5eee9226-60f3-40dc-a8cd-6d10f46ea53b)

xmnlab commented:
thanks @apkrelling , for now I don't have any objections, maybe it something to discuss in the future, but for now it is ok. thanks, appreciate that!
